src/main/java/com/aws/greengrass/authorization/AuthorizationModule.java

```diff
@@ -42,17 +52,57 @@ public void addPermission(String destination, Permission permission) throws Auth
                 || Utils.isEmpty(permission.getOperation())) {
             throw new AuthorizationException("Invalid arguments");
         }
-        // resource as null is ok, but it should not be empty
         String resource = permission.getResource();
-        if (resource != null && Utils.isEmpty(resource)) {
-            throw new AuthorizationException("Resource cannot be empty");
-        }
+        String validatedResource = validateResource(resource);
         resourceAuthZCompleteMap.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                resource);
         rawResourceList.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                validatedResource);
+    }
+
+    /**
+     * Only allow '?' if it's escaped. You can only escape special characters ('*', '$', '?').
+     * Any occurrence of '${' is only valid if it holds a single valid special character ('*', '$', '?') inside it
+     * and ends with '}'. (eg: "${*}" is valid, "${c}" is invalid, "${c" is invalid, ${*bc} is invalid)
+     * @param resource resource to be parsed
+     */
+    private String validateResource(String resource) throws AuthorizationException {
+        if (resource == null) {
+            return null;
+        }
+        // resource as null is ok, but it should not be empty
+        if (Utils.isEmpty(resource)) {
+            throw new AuthorizationException("Resource cannot be empty");
+        }
+        int length = resource.length();
+        StringBuilder sb = new StringBuilder(length);
+        for (int i = 0; i < length; i++) {
+            char currentChar = resource.charAt(i);
+            if (currentChar == escapeChar && i + 1 < length && resource.charAt(i + 1) == '{') {
+                char actualChar = getActualChar(resource.substring(i));
+                if (actualChar == nullChar) {
+                    throw new AuthorizationException("Resource not allowed, incorrect escape sequence used");
+                }
+                if (!specialChars.contains(actualChar)) {
```

Comment:
if this is the only place it's being used, why not make a method like:

```
boolean isSpecialChar(char actualChar) {
  return actualChar == wildcardChar || actualChar == escapeChar || actualChar == singleCharWildcard;
}
```

There's only 3 characters to check - no need create an array list and store that.

Reply:
Yes, I can make a method for this. Thanks